### ISSUE TITLE
Problems with MaskablePPO

### ISSUE THREAD

**Issue by koliber31** [Stable-Baselines-Team/stable-baselines3-contrib#195]
### 🐛 Bug

Hi
I had problems with maskable ppo which I described here https://github.com/DLR-RM/stable-baselines3/issues/1596. I thought that I found solution in one of issues https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/81#issuecomment-1179910351. The problem is that error stopped occuring but in the same time agent lost its ability to learn. Below are screenshots of mean rewards, 150k and 260k timesteps are for case with error and 4M timesteps is for case without error. 
![150kSteps](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/assets/42122590/9be94782-c74e-43c7-a22f-bb1c95f24d6a)
![260kSteps](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/assets/42122590/ca19c616-3921-42fe-9b51-74949d4dca69)
![wykresy](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/assets/42122590/ea241d05-b1f0-4659-a54e-e3a9ecd28ba2)
Unfortunately I don't have screenshots of learning process where agent menaged to get mean reward of ~-0.75 before error.

### Code example

The only thing I changed since last issue in code is solution from #81 
```
# Reinitialize with updated logits
        super().__init__(logits=logits, validate_args=False)

        # self.probs may already be cached, so we must force an update
        self.probs = logits_to_probs(self.logits)
```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues) in the repo
- [X] I have read the [documentation](https://sb3-contrib.readthedocs.io/en/master/)
- [X] I have provided a [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014) example to reproduce the bug
- [X] I have checked my env using the env checker
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by yiptsangkin** [comment]
i try to modify the source code change validate_args to False and it works

**Comment by yiptsangkin** [comment]
dtype also have problem sometimes

**Comment by koliber31** [comment]
> i try to modify the source code change validate_args to False and it works

The error stops occuring in my case too but it doesn't learn with `validate_args=False`.
Does it in your case? Could you provide any screen of learning curve? 
What is dtype in your case? I tried int32, float32 and none didn't help.

**Comment by yiptsangkin** [comment]
same problem, i think this need some tech support to solve this problem.

**Comment by yiptsangkin** [comment]
https://discuss.pytorch.org/t/distributions-categorical-fails-with-constraint-simplex-but-manual-check-passes/163209/6

**Comment by yiptsangkin** [comment]
https://github.com/pytorch/pytorch/issues/87468

**Comment by koliber31** [comment]
> [pytorch/pytorch#87468](https://github.com/pytorch/pytorch/issues/87468)

Did you menage to solve problem using this solution from this issue?

**Comment by yiptsangkin** [comment]
> 

yes，at least my env works！

**Comment by yiptsangkin** [comment]
> > [pytorch/pytorch#87468](https://github.com/pytorch/pytorch/issues/87468)
> 
> Did you menage to solve problem using this solution from this issue?

  class _Simplex(Constraint):
     ...
      def check(self, value):
-         return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < 1e-6)
+         tol = torch.finfo(value.dtype).eps * 10 * value.size(-1) ** 0.5
+         return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < tol)

**Comment by koliber31** [comment]
> 

It looks like this in my case 
```
class _Simplex(Constraint):
    """
    Constrain to the unit simplex in the innermost (rightmost) dimension.
    Specifically: `x >= 0` and `x.sum(-1) == 1`.
    """
    event_dim = 1

    def check(self, value):
        # return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < 1e-6)
        tol = torch.finfo(value.dtype).eps * 10 * value.size(-1) ** 0.5
        return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < tol)
```
And my mask doesn't work after this change. Agent keeps making invalid moves.

**Comment by yiptsangkin** [comment]
> > 
> 
> It looks like this in my case
> 
> ```
> class _Simplex(Constraint):
>     """
>     Constrain to the unit simplex in the innermost (rightmost) dimension.
>     Specifically: `x >= 0` and `x.sum(-1) == 1`.
>     """
>     event_dim = 1
> 
>     def check(self, value):
>         # return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < 1e-6)
>         tol = torch.finfo(value.dtype).eps * 10 * value.size(-1) ** 0.5
>         return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < tol)
> ```
> 
> And my mask doesn't work after this change. Agent keeps making invalid moves. Edit: It keeps making invalid moves even without this change. Had to make some mistake in between :(. Does this constraint look like this in your case and does it learn?

this is the source code of torch, you can see the answer https://github.com/pytorch/pytorch/issues/87468#issuecomment-1294411284, edit the source code and it works. my env can learn without errors and the performance is same as before this change.

**Comment by yiptsangkin** [comment]
> > 
> 
> It looks like this in my case
> 
> ```
> class _Simplex(Constraint):
>     """
>     Constrain to the unit simplex in the innermost (rightmost) dimension.
>     Specifically: `x >= 0` and `x.sum(-1) == 1`.
>     """
>     event_dim = 1
> 
>     def check(self, value):
>         # return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < 1e-6)
>         tol = torch.finfo(value.dtype).eps * 10 * value.size(-1) ** 0.5
>         return torch.all(value >= 0, dim=-1) & ((value.sum(-1) - 1).abs() < tol)
> ```
> 
> And my mask doesn't work after this change. Agent keeps making invalid moves. Edit: It keeps making invalid moves even without this change. Had to make some mistake in between :(. Does this constraint look like this in your case and does it learn?

if the agent take error action, maybe you should check you mask of action.

**Comment by koliber31** [comment]
> 

I did change source code and this is what it looks like after change. You are right my action mask had error. Now its seems to be learning but rewards is increasing slowly. And i still have problem that mean length of episode is increasing even though I know can win fast in some cases. I'll try to give feedback if it menaged to learn.

**Comment by yiptsangkin** [comment]
> > 
> 
> I did change source code and this is what it looks like after change. You are right my action mask had error. Now its seems to be learning but rewards is increasing slowly. And i still have problem that mean length of episode is increasing even though I know can win fast in some cases. I'll try to give feedback if it menaged to learn.

if the rewards increasing slowly, maybe is the problem of reward function.

**Comment by koliber31** [comment]
> > > 
> > 
> > 
> > I did change source code and this is what it looks like after change. You are right my action mask had error. Now its seems to be learning but rewards is increasing slowly. And i still have problem that mean length of episode is increasing even though I know can win fast in some cases. I'll try to give feedback if it menaged to learn.
> 
> if the rewards increasing slowly, maybe is the problem of reward function.

My rewards are just:
win = 1
lose = -1
draw = 0.25
Does your changed source code looks like mine? And did you change class _Symmetric too or just _Simplex?
